### ISSUE TITLE
Lock contention when low-level OS interaction(s) are stuck

### ISSUE THREAD

**Issue by estesp** [docker/docker#22671]
Data below, but bringing two (IMO) important questions to the top:

 1. Should we consider whether locking (at the daemon container object level) during a potentially long-lived operation impacts daemon reliability? (hint: answer is probably "it depends")
 2. If we decide locking is critically necessary through the operation, is there a way to timebox low-level operations or determine viability of a long-running call at kernel or lower OS layers?

*Situation:*
A container is in the state of natural exit/teardown and as part of the normal course of the **monitor**'s cleanup operations, the libnetwork endpoint teardown occurred. In this case, the netlink interactions with the kernel seem to have hung on a `recvfrom()` call waiting on a response from a netlink message. Apparently, the netlink response never happened/never arrived, because a **very long time** has passed and this thread is still blocked. As it happens, this thread is holding the container's lock at this point due to code in `container/monitor.go` (from 1.10.3; but 1.11.x has similar code holding the same lock in `daemon/monitor.go`):
```
151     // ensure that when the monitor finally exits we release the networking and unmount the rootfs
152     defer func() {
153         if afterRun {
154             m.container.Lock()
155             defer m.container.Unlock()
156             m.container.SetStopped(&exitStatus)
157         }
158         m.Close()
159     }()
```

A netlink hiccup is not the only potential problem here as a poorly configured devicemapper graphdriver can also present a similar scenario on delete, although it rarely blocks forever, but it can be minutes under extremely heavy load.

*Note that the deferred `m.container.Unlock()` will not happen until after `m.Close()` returns, which won't happen in this case (stacktrace below) because of the netlink hang.*

```
goroutine 167855627 [syscall, 1285 minutes, locked to thread]:
     syscall.Syscall6(0x2d, 0x1c9, 0xc82808c000, 0x1000, 0x0, 0xc82b618b40, 0xc82b618b3c, 0xc82002f500, 0xc832d05980, 0x5c05cc)
         /usr/local/go/src/syscall/asm_linux_amd64.s:44 +0x5
     syscall.recvfrom(0x1c9, 0xc82808c000, 0x1000, 0x1000, 0x0, 0xc82b618b40, 0xc82b618b3c, 0x423cc9, 0x0, 0x0)
         /usr/local/go/src/syscall/zsyscall_linux_amd64.go:1712 +0x91
     syscall.Recvfrom(0x1c9, 0xc82808c000, 0x1000, 0x1000, 0x0, 0x1000, 0x0, 0x0, 0x0, 0x0)
         /usr/local/go/src/syscall/syscall_unix.go:244 +0xb8
     github.com/vishvananda/netlink/nl.(*NetlinkSocket).Receive(0xc835f1d0e0, 0x0, 0x0, 0x0, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/vishvananda/netlink/nl/nl_linux.go:341 +0xae
     github.com/vishvananda/netlink/nl.(*NetlinkRequest).Execute(0xc8362d1c80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/vishvananda/netlink/nl/nl_linux.go:228 +0x1ff
     github.com/vishvananda/netlink.LinkSetName(0x7fbcd40c77f8, 0xc83346e240, 0xc84df36f50, 0xd, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/vishvananda/netlink/link_linux.go:105 +0x34d
     github.com/docker/libnetwork/osl.(*nwIface).Remove.func2(0x8, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/osl/interface_linux.go:129 +0x155
     github.com/docker/libnetwork/osl.nsInvoke(0xc82d491530, 0x22, 0x194d760, 0xc82b6191e0, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/osl/namespace_linux.go:295 +0x46a
     github.com/docker/libnetwork/osl.(*nwIface).Remove(0xc854344900, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/osl/interface_linux.go:158 +0x12d  
     github.com/docker/libnetwork.releaseOSSboxResources(0x7fbcd40c82c0, 0xc854547680, 0xc8554a2a00)
         /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:591 +0x17f      
     github.com/docker/libnetwork.(*sandbox).clearNetworkResources(0xc8368e2000, 0xc84fbab200, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:704 +0x256  
     github.com/docker/libnetwork.(*endpoint).sbLeave(0xc84fbab200, 0x7fbcd40c7d00, 0xc8368e2000, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/endpoint.go:586 +0x8ce
     github.com/docker/libnetwork.(*endpoint).Leave(0xc8554a2a00, 0x7fbcd40c7d00, 0xc8368e2000, 0x0, 0x0, 0x0, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/endpoint.go:536 +0x1bf
     github.com/docker/libnetwork.(*sandbox).delete(0xc8368e2000, 0xc8351cb800, 0x0, 0x0)           
         /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:199 +0xad4
     github.com/docker/libnetwork.(*sandbox).Delete(0xc8368e2000, 0x0, 0x0)
         /go/src/github.com/docker/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:166 +0x32    
     github.com/docker/docker/daemon.(*Daemon).releaseNetwork(0xc820660780, 0xc822961a40)        
         /go/src/github.com/docker/docker/daemon/container_operations_unix.go:1013 +0x4f1      
     github.com/docker/docker/daemon.(*Daemon).Cleanup(0xc820660780, 0xc822961a40) 
         /go/src/github.com/docker/docker/daemon/start.go:159 +0x3d           
     github.com/docker/docker/container.(*containerMonitor).Close(0xc841337f00, 0x0, 0x0)
         /go/src/github.com/docker/docker/container/monitor.go:127 +0x68
     github.com/docker/docker/container.(*containerMonitor).start.func1(0xc82b619d5f, 0xc841337f00, 0xc82b619d60)

```

The problem at this point is that the container's lock is held (indefinitely in this case) and will start to have immediate impact on the rest of the daemon. Today, a `docker ps` or even `docker inspect` will start to hang based on the need to acquire the same container's lock:
```
goroutine 167876743 [semacquire, 1285 minutes]:
     sync.runtime_Semacquire(0xc824b92c44)
         /usr/local/go/src/runtime/sema.go:43 +0x26
     sync.(*Mutex).Lock(0xc824b92c40)
         /usr/local/go/src/sync/mutex.go:82 +0x1c4
     github.com/docker/docker/daemon.(*Daemon).reducePsContainer(0xc820660780, 0xc822961a40, 0xc82e1daf50, 0xc83e913430, 0x0, 0x0, 0x0)
         /go/src/github.com/docker/docker/daemon/list.go:126 +0x49
     github.com/docker/docker/daemon.(*Daemon).reduceContainers(0xc820660780, 0xc84647f3b0, 0xc83e913430, 0x0, 0x0, 0x0, 0x0, 0x0)
         /go/src/github.com/docker/docker/daemon/list.go:109 +0x1a0
     github.com/docker/docker/daemon.(*Daemon).Containers(0xc820660780, 0xc84647f3b0, 0x0, 0x0, 0x0, 0x0, 0x0)
         /go/src/github.com/docker/docker/daemon/list.go:96 +0x6a
     github.com/docker/docker/api/server/router/container.(*containerRouter).getContainersJSON(0xc821239980, 0x7fbcd40c6b78, 0xc83a590930, 0x7fbcd40c6b40, 0xc83b908bb0, 0xc8368bc380, 0xc83a5908d0, 0x0, 0x0)
         /go/src/github.com/docker/docker/api/server/router/container/container_routes.go:51 +0x379
```

Things will get worse when there is an interaction between the lock for the `memoryStore` and this blocked container's lock; specifically if anything requires the full "rw" lock on `memoryStore` (and gets it) and then blocks on the container's held lock, the daemon at that point will be fully "hung" from an API standpoint. In the below goroutine, the culprit who starts this is a `docker rmi` of the image that was actually in use by the container which was in the process of exit/teardown (whose lock is currently held, in case anyone forgot in this long story).  Note that the `memoryStore` lock is held because of the walk via `First()`, as it tries to make sure there are no active containers using the image, but now this image delete is going to block forever on the container lock to check `IsRunning()` status.

```
goroutine 168199920 [semacquire, 430 minutes]:
    sync.runtime_Semacquire(0xc824b92c44)
        /usr/local/go/src/runtime/sema.go:43 +0x26
    sync.(*Mutex).Lock(0xc824b92c40)
        /usr/local/go/src/sync/mutex.go:82 +0x1c4                                                                                                                 
    github.com/docker/docker/container.(*State).IsRunning(0xc824b92c40, 0x1182040)            
        /go/src/github.com/docker/docker/container/state.go:161 +0x28
    github.com/docker/docker/daemon.(*Daemon).checkImageDeleteConflict.func1(0xc822961a40, 0xc8210cde00)
        /go/src/github.com/docker/docker/daemon/image_delete.go:328 +0x3e  
    github.com/docker/docker/container.(*memoryStore).First(0xc820ca2280, 0xc842556f80, 0x0)  
        /go/src/github.com/docker/docker/container/memory_store.go:66 +0xf2
    github.com/docker/docker/daemon.(*Daemon).checkImageDeleteConflict(0xc820660780, 0xc823a245f0, 0x47, 0xf, 0x2748d899) 
        /go/src/github.com/docker/docker/daemon/image_delete.go:330 +0x1c8 
    github.com/docker/docker/daemon.(*Daemon).imageDeleteHelper(0xc820660780, 0xc823a245f0, 0x47, 0xc83dc3b430, 0x10100, 0x0, 0x0)  
        /go/src/github.com/docker/docker/daemon/image_delete.go:263 +0x79
    github.com/docker/docker/daemon.(*Daemon).ImageDelete(0xc820660780, 0xc8255719b5, 0x29, 0x100, 0x0, 0x0, 0x0, 0x0, 0x0) 
        /go/src/github.com/docker/docker/daemon/image_delete.go:162 +0xd1a   
    github.com/docker/docker/api/server/router/local.(*router).deleteImages(0xc820338020, 0x7fbcd40c6b78, 0xc84aa66f90, 0x7fbcd40c6b40, 0xc830a3b290, 0xc8214c09a0, 0xc84aa66e40, 0x0, 0x0)
```

At this point all hope is lost for this daemon (unless the netlink call miraculously completes), as any container interactions (create/start, stop, rm) will need the `memoryStore` lock, which is currently held by a blocked thread.

My curiosity at the simplest "fix" is whether the container lock needs to be held during the `m.Close()` operation as the lock seems most useful for state changes or metadata protection, which will not happen during the lower-level operations.

**Comment by LK4D4** [comment]
@aboch do you think it's possible for us to add a timeout to Netlink operations? I see that now it's possible only with `unsafe`, but I mean if we change the code.

**Comment by cpuguy83** [comment]
I wonder too if we can have shorter lived locks for simpler tasks like reading/writing to the container struct vs waiting on something to happen.
This would allow things like inspect/ps to not be blocked by longer operations since it can just grab a copy of the struct and get out.

**Comment by justincormack** [comment]
I believe that the netlink never returning is a kernel bug, although I can't find the issue for it right now, maybe @thaJeztah can as he seems to be able to find all issues. Netlink is not guaranteed to be reliable, but not sure if this manifests as an error reply or a timeout (it is very hard to replicate).

**Comment by thaJeztah** [comment]
@justincormack possibly you mean https://github.com/docker/docker/issues/5618 ?

**Comment by thaJeztah** [comment]
Let me close this ticket for now, as it looks like it went stale.
